### PR TITLE
Configurable retry strategy

### PR DESCRIPTION
**Pull request by lchanmann**
## Ref
- https://nanopay.atlassian.net/browse/NP-6077

## Changes
- Add RetryStrategy interface
- Implement DefaultRetryStrategy and ExponentialBackoffRetryStrategy
- Update Rule and RetryManager to support retry strategy config

### CODE REVIEW

**Comment by jlhughes** [comment]
Thinking
package: foam.util.retry 
Interface 
long maxRetries  0 - infinite
long getDelay(x)
boolean canRetry(x) or getDelay(x) throws RetryLimitReachedException (or similar)

SimpleRetryStrategy
BackoffRetryStrategy
.... 
Will be used by Medusa, SAF, Rules
See similar in SAF https://github.com/kgrgreer/foam3/pull/650

File: src/foam/nanos/ruler/RetryManager.java [hunk 1]
```diff
@@ -51,19 +50,14 @@ public void run() {
...
-
-  public RetryManager(int maxRetry, int retryDelay, String description) {
-    maxRetry_ = maxRetry;
-    retryDelay_ = retryDelay;
-    description_ = description;
-    latch_ = new CountDownLatch(maxRetry);
+  public RetryManager(RetryStrategy retryStrategy, String description) {
+    retryStrategy_ = retryStrategy;
+    description_   = description;
+    latch_         = new CountDownLatch(retryStrategy_.getMaxRetry());
```
Comment (jlhughes) [inline src/foam/nanos/ruler/RetryManager.java:67]: The catch interrupt below - should cancel the retry - An interrupt generally only occurs when the system is shutting down, so don't want this blocking.

**Review by jlhughes — COMMENTED** [review]
Interesting approach. 
How would we support retry forever - a ForeverRetryStrategy? 
And RetryUntilStrategy could take a date/time.  or any predicate.

File: src/foam/nanos/ruler/RetryManager.java (hunk 1 above)
Comment (lchanmann) [reply src/foam/nanos/ruler/RetryManager.java:67]: Fixed in https://github.com/kgrgreer/foam3/pull/1223/commits/cde7ef485bd77b2af8316d87b049b0896c9f754a.

**Review by jlhughes — COMMENTED** [review]
LGTM

**Review by nanoscott — APPROVED** [review]

**Pull request merged**